File: 2026/03- March/12-Thursday/solution.py
from typing import List


class Solution:
    def maxStability(self, n: int, edges: List[List[int]], k: int) -> int:
        parent = list(range(n))
        rank = [0] * n

        def find(x):
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x

        def union(a, b):
            ra, rb = find(a), find(b)
            if ra == rb:
                return False
            if rank[ra] < rank[rb]:
                parent[ra] = rb
            elif rank[ra] > rank[rb]:
                parent[rb] = ra
            else:
                parent[rb] = ra
                rank[ra] += 1
            return True

        dsu_parent = list(range(n))
        dsu_rank = [0] * n

        def find0(x):
            while dsu_parent[x] != x:
                dsu_parent[x] = dsu_parent[dsu_parent[x]]
                x = dsu_parent[x]
            return x

        def union0(a, b):
            ra, rb = find0(a), find0(b)
            if ra == rb:
                return False
            if dsu_rank[ra] < dsu_rank[rb]:
                dsu_parent[ra] = rb
            elif dsu_rank[ra] > dsu_rank[rb]:
                dsu_parent[rb] = ra
            else:
                dsu_parent[rb] = ra
                dsu_rank[ra] += 1
            return True

        mandatory_count = 0
        for u, v, s, m in edges:
            if m == 1:
                if not union0(u, v):
                    return -1
                mandatory_count += 1

        if mandatory_count > n - 1:
            return -1

        def can(x):
            parent = list(range(n))
            rank = [0] * n

            def find(x):
                while parent[x] != x:
                    parent[x] = parent[parent[x]]
                    x = parent[x]
                return x

            def union(a, b):
                ra, rb = find(a), find(b)
                if ra == rb:
                    return False
                if rank[ra] < rank[rb]:
                    parent[ra] = rb
                elif rank[ra] > rank[rb]:
                    parent[rb] = ra
                else:
                    parent[rb] = ra
                    rank[ra] += 1
                return True

            upgrades = 0
            edges_used = 0

            for u, v, s, m in edges:
                if m == 1:
                    if s < x:
                        return False
                    if union(u, v):
                        edges_used += 1

            optional = []

            for u, v, s, m in edges:
                if m == 0:
                    if s >= x:
                        optional.append((0, u, v))
                    elif s * 2 >= x:
                        optional.append((1, u, v))

            optional.sort()

            for cost, u, v in optional:
                if union(u, v):
                    edges_used += 1
                    upgrades += cost

                    if upgrades > k:
                        return False

                    if edges_used == n - 1:
                        return True

            return edges_used == n - 1 and upgrades <= k

        lo, hi = 0, 2 * 10**5
        ans = -1

        while lo <= hi:
            mid = (lo + hi) // 2

            if can(mid):
                ans = mid
                lo = mid + 1
            else:
                hi = mid - 1

        return ans

File: 2026/03- March/12-Thursday/solution.java
import java.util.*;

class Solution {

    public int maxStability(int n, int[][] edges, int k) {

        int[] dsuParent = new int[n];
        int[] dsuRank = new int[n];

        for (int i = 0; i < n; i++) {
            dsuParent[i] = i;
        }

        int mandatoryCount = 0;

        for (int[] e : edges) {
            int u = e[0], v = e[1], s = e[2], m = e[3];

            if (m == 1) {
                if (!union0(u, v, dsuParent, dsuRank)) {
                    return -1;
                }
                mandatoryCount++;
            }
        }

        if (mandatoryCount > n - 1) {
            return -1;
        }

        int lo = 0, hi = 200000;
        int ans = -1;

        while (lo <= hi) {
            int mid = (lo + hi) / 2;

            if (can(mid, n, edges, k)) {
                ans = mid;
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }

        return ans;
    }

    private boolean can(int x, int n, int[][] edges, int k) {

        int[] parent = new int[n];
        int[] rank = new int[n];

        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }

        int upgrades = 0;
        int edgesUsed = 0;

        for (int[] e : edges) {
            int u = e[0], v = e[1], s = e[2], m = e[3];

            if (m == 1) {
                if (s < x) return false;

                if (union(u, v, parent, rank)) {
                    edgesUsed++;
                }
            }
        }

        List<int[]> optional = new ArrayList<>();

        for (int[] e : edges) {
            int u = e[0], v = e[1], s = e[2], m = e[3];

            if (m == 0) {
                if (s >= x) {
                    optional.add(new int[]{0, u, v});
                } else if (s * 2 >= x) {
                    optional.add(new int[]{1, u, v});
                }
            }
        }

        optional.sort(Comparator.comparingInt(a -> a[0]));

        for (int[] e : optional) {
            int cost = e[0], u = e[1], v = e[2];

            if (union(u, v, parent, rank)) {
                edgesUsed++;
                upgrades += cost;

                if (upgrades > k) return false;

                if (edgesUsed == n - 1) return true;
            }
        }

        return edgesUsed == n - 1 && upgrades <= k;
    }

    private int find(int x, int[] parent) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }

    private boolean union(int a, int b, int[] parent, int[] rank) {

        int ra = find(a, parent);
        int rb = find(b, parent);

        if (ra == rb) return false;

        if (rank[ra] < rank[rb]) {
            parent[ra] = rb;
        } else if (rank[ra] > rank[rb]) {
            parent[rb] = ra;
        } else {
            parent[rb] = ra;
            rank[ra]++;
        }

        return true;
    }

    private int find0(int x, int[] parent) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }

    private boolean union0(int a, int b, int[] parent, int[] rank) {

        int ra = find0(a, parent);
        int rb = find0(b, parent);

        if (ra == rb) return false;

        if (rank[ra] < rank[rb]) {
            parent[ra] = rb;
        } else if (rank[ra] > rank[rb]) {
            parent[rb] = ra;
        } else {
            parent[rb] = ra;
            rank[ra]++;
        }

        return true;
    }
}
